py_ecc/fields/field_elements.py

```diff
@@ -0,0 +1,377 @@
+from typing import (
+    cast,
+    List,
+    Sequence,
+    Union,
+)
+
+from py_ecc.utils import (
+    deg,
+    poly_rounded_div,
+    prime_field_inv,
+)
+
+
+IntOrFQ = Union[int, "FQ"]
+
+
+class FQ(object):
+    """
+    A class for field elements in FQ. Wrap a number in this class,
+    and it becomes a field element.
+    """
+    n = None  # type: int
+    field_modulus = None
+    # curve_name can be either 'bn128' or 'bls12_381'
+    # This is needed to obtain field_modulus, FQ2_MODULUS_COEFFS
+    # and FQ12_MODULUS_COEFFS from the curve properties
+    curve_name = None
+
+    def __init__(self, val: IntOrFQ) -> None:
+        if self.curve_name is None:
+            raise AttributeError("Curve Name hasn't been specified")
+        if self.field_modulus is None:
+            raise AttributeError("Field Modulus hasn't been specified")
+
+        if isinstance(val, FQ):
+            self.n = val.n
+        elif isinstance(val, int):
+            self.n = val % self.field_modulus
+        else:
+            raise TypeError(
+                "Expected an int or FQ object, but got object of type {}"
+                .format(type(val))
+            )
+
+    def __add__(self, other: IntOrFQ) -> "FQ":
+        if isinstance(other, FQ):
+            on = other.n
+        elif isinstance(other, int):
```

Comment:
I don't know if this should be experimented with before merging this PR or something we try after but... what if we just made `FQ` a subclass of `int`.  My intuition says it could solve a whole swath of these various type checks that occur since we could treat `FQ` instances as integers.

What isn't clear to me is how difficult or necessary it would be to ensure that an `FQ` object is what is return returned from operations like `a * b` and `b * a` where one of `a, b` is an `FQ` and one is an `int`.  Stated differently, it seems potentially problematic if it is difficult to reason about whether an object we are interacting with is an integer or an `FQ` as well as whether the return value of various mathmatical operations will be an `int` or `FQ`

Reply:
@pipermerriam I see what you are saying. It should work as per theory. But I would like this experimentation to be done after the refactoring. Opened https://github.com/ethereum/py_ecc/issues/54 for this.